File: SynthesisMNTE/Python/bounds.py
#####################################################################################################################
# Accounting for Missing Data in Public Health Research Using a Synthesis of Statistical and Mathematical Models
#   This file runs the bounds and sensitivity analysis reported in the main paper.
#
# Paul Zivich (2025/10/03)
#####################################################################################################################

###############################################
# Loading packages

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from delicatessen import MEstimator
from delicatessen.estimating_equations import ee_regression

np.random.seed(905141)

###############################################
# Setting up data

d = pd.read_csv("../data/nhanes.csv")
d = d.dropna(subset=['height', ])
d['height'] = d['height'] / 2.54
d['intercept'] = 1

###############################################
# Synthesis bounds

reg_cols = []  # Storage for new regression column names
for j in [8, 9, 10, 11, 12, 13, 14, 15, 16, 17]:  # For each unique age in the positive region
    age_label = 'age' + str(j)  # ... create a new column label
    reg_cols.append(age_label)  # ... add column label to the running list
    d[age_label] = np.where(d['age'] == j, 1, 0)  # ... create indicator variable for that specific age


def psi(theta):
    mu = theta[0]
    beta = theta[1:]

    # Data preparation for missing
    r = np.where(d['sbp'].isna(), 0, 1)                     # Indicator if the outcome was observed
    y_no_miss = np.where(r == 1, d['sbp'], -9999)           # Replacing NaN with -9999 to avoid NaN-related errors
    x_star = np.where(d['age'] < 8, 0, 1)

    # Estimating function for statistical nuisance model
    ee_reg = ee_regression(beta,                            # Estimating functions for a regression model of
                           X=d[reg_cols],                   # ... indicator terms for age
                           y=y_no_miss,                     # ... on outcomes where NaN was replaced
                           model='linear',                  # ... using linear regression
                           weights=d['sample_weight']) * r  # ... using sample weights and among complete-cases only

    # Estimating function for bound
    yhat = np.dot(d[reg_cols], beta)                        # Predicting SBP for all observations
    yhat = np.where(x_star == 1, yhat, mu_nonpos)           # Setting SBP for those in nonpositive region
    ee_mean = d['sample_weight']*(yhat - mu)                # Compute the estimating function for the weighted mean

    return np.vstack([ee_mean, ee_reg])                     # Returning stacked estimating equations


# Computing lower bound with M-estimator
mu_nonpos = 70
estr = MEstimator(psi, init=[100., ] + [100., ] * len(reg_cols))
estr.estimate()
lbound = estr.theta[0]
lbound_ci = estr.confidence_intervals()[0, 0]

# Computing upper bound with M-estimator
mu_nonpos = 120
estr = MEstimator(psi, init=[100., ] + [100., ] * len(reg_cols))
estr.estimate()
ubound = estr.theta[0]
ubound_ci = estr.confidence_intervals()[0, 1]

# Results
print("Synthesis Bounds")
print("Bounds:", [lbound, ubound])
print("95% CI:", [lbound_ci, ubound_ci])

###############################################
# Synthesis sensitivity analysis

# stat-model parameters
beta_hat = estr.theta[1:]

p, l, u = [], [], []
values = np.linspace(70, 120, 40)

for v in values:
    # Fitting estimator for given point
    mu_nonpos = v
    estr = MEstimator(psi, init=[100., ] + list(beta_hat))
    estr.estimate()
    ci = estr.confidence_intervals()[0, :]

    # Storing output results
    p.append(estr.theta[0])
    l.append(ci[0])
    u.append(ci[1])


# Generating figure
plt.axhline(104.7, color='#298c8c', label='Complete-case', zorder=1)
plt.fill_between([70, 120], [104.1, 104.1], [105.3, 105.3], color='#298c8c', alpha=0.2, zorder=2)
plt.axhline(101.6, color='#800074', label='Extrapolation', zorder=3)
plt.fill_between([70, 120], [100.8, 100.8], [102.4, 102.4], color='#800074', alpha=0.2, zorder=4)
plt.plot(values, p, '-', color='k', label='Sensitivity Analysis', zorder=6)
plt.fill_between(values, l, u, color='k', alpha=0.2, zorder=5)
plt.xlabel(r"Input Values for $E[Y \mid X^* = 0]$")
plt.ylabel("Mean Systolic Blood Pressure")
plt.ylim([90, 115])
plt.xlim([70, 120])
plt.legend()
plt.tight_layout()
plt.show()

File: SynthesisMNTE/R/bounds.R
#####################################################################################################################
# Accounting for Missing Data in Public Health Research Using a Synthesis of Statistical and Mathematical Models
#   This file runs the bounds reported in the main paper.
#
# Paul Zivich (2025/10/14)
#####################################################################################################################

library(plyr)
library(geex)
library(resample)
library(dplyr)


###############################################
# Setting up data

setwd("C:/Users/zivic/Documents/open-source/publications-code/SynthesisMNTE")
nhanes <- read.csv("data/nhanes.csv")
nhanes$agelt8 <- ifelse(nhanes$age<8, 1, 0)

nhanes$age8 <- ifelse(nhanes$age == 8, 1, 0)
nhanes$age9 <- ifelse(nhanes$age == 9, 1, 0)
nhanes$age10 <- ifelse(nhanes$age == 10, 1, 0)
nhanes$age11 <- ifelse(nhanes$age == 11, 1, 0)
nhanes$age12 <- ifelse(nhanes$age == 12, 1, 0)
nhanes$age13 <- ifelse(nhanes$age == 13, 1, 0)
nhanes$age14 <- ifelse(nhanes$age == 14, 1, 0)
nhanes$age15 <- ifelse(nhanes$age == 15, 1, 0)
nhanes$age16 <- ifelse(nhanes$age == 16, 1, 0)
nhanes$age17 <- ifelse(nhanes$age == 17, 1, 0)


###############################################
# Synthesis bounds

#to avoid missing values in dataset, set all missing values to 999
nhanes.hold <- nhanes
nhanes.hold$sbp <- ifelse(nhanes$miss==1, 999, nhanes$sbp)


#estimate the variance using M-estimation (geex package)
estfun_extrap <- function(data, models){
  W <- data$sample_weight
  M <- data$miss
  Y <- data$sbp

  #fit outcome model (by missingness variable, so we can later limit to complete case)
  Xmat <- grab_design_matrix(data=data, rhs_formula=grab_fixed_formula(models$out))
  out_scores <- grab_psiFUN(models$out, data)
  out_pos <- 1:ncol(Xmat)

  #     yhat = np.where(x_star == 1, yhat, mu_nonpos)           # Setting SBP for those in nonpositive region

  function(theta){
      p <- length(theta)
      #get predicted values from model for everyone
      Y.imp <- Xmat %*% theta[out_pos]
      Y.imp <- ifelse(data$agelt8 == 1, boundary, Y.imp)
      
      #estimating equations
      c(W*out_scores(theta[out_pos])*(1-M),
        W*(Y.imp-theta[p]))
  }
}

#function to call M-estimator
geex_extrap <- function(data, out_formula){
  out_model  <- glm(out_formula, data=data)
  models <- list(out=out_model)

  geex_results_extrap <- m_estimate(
    estFUN = estfun_extrap,
    data = data,
    root_control = setup_root_control(start=c(coef(out_model), 100)),
    outer_args = list(models = models))
  return(geex_results_extrap)
}

### Lower bound ###
boundary <- 70
bound.res <- geex_extrap(nhanes.hold, sbp ~ age8 + age9 + age10 + age11 + age12
                         + age13 + age14 + age15 + age16 + age17 - 1)

#format output and compute CIs
bound.est <- bound.res@estimates[length(bound.res@estimates)]
bound.se <- as.numeric(sqrt(bound.res@vcov[length(bound.res@estimates),length(bound.res@estimates)]))
bound <- as.data.frame(cbind(bound.est,bound.se))
names(bound) <- c('est','se')
bound$LCL <- as.numeric(bound$est) - 1.96*as.numeric(bound$se)
bound$type <- "Lower"
bound

### Upper bound ###
boundary <- 120
bound.res <- geex_extrap(nhanes.hold, sbp ~ age8 + age9 + age10 + age11 + age12
                         + age13 + age14 + age15 + age16 + age17 - 1)

#format output and compute CIs
bound.est <- bound.res@estimates[length(bound.res@estimates)]
bound.se <- as.numeric(sqrt(bound.res@vcov[length(bound.res@estimates),length(bound.res@estimates)]))
bound <- as.data.frame(cbind(bound.est,bound.se))
names(bound) <- c('est','se')
bound$UCL <- as.numeric(bound$est) + 1.96*as.numeric(bound$se)
bound$type <- "Upper"
bound


# Output
#       est        se      LCL  type
#  92.70003 0.4292731 91.85866 Lower
# 
#     est        se     UCL  type
#  109.93 0.2739797 110.467 Upper
